Specify underlying type for enum classes in CryptoKey

szewai opened the pull request:
#### bf1655dee5e2866f57e28ef03d04cbc736abfabc
<pre>
Specify underlying type for enum classes in CryptoKey
<a href="https://bugs.webkit.org/show_bug.cgi?id=283381">https://bugs.webkit.org/show_bug.cgi?id=283381</a>
<a href="https://rdar.apple.com/140231507">rdar://140231507</a>

Reviewed by Per Arne Vollan.

Also, remove paramter in isPlatformSupportedCurve() and rename it since the return value does not rely on the NamedCurve
value.

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
* Source/WebCore/crypto/CryptoKey.h:
* Source/WebCore/crypto/CryptoKeyUsage.h:
* Source/WebCore/crypto/SubtleCrypto.h:
* Source/WebCore/crypto/cocoa/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::supportsNamedCurve):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::isPlatformSupportedCurve): Deleted.
* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::CryptoKeyOKP::supportsNamedCurve):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::isPlatformSupportedCurve): Deleted.
* Source/WebCore/crypto/keys/CryptoKeyEC.h:
* Source/WebCore/crypto/keys/CryptoKeyOKP.cpp:
(WebCore::CryptoKeyOKP::generatePair):
(WebCore::CryptoKeyOKP::importRaw):
(WebCore::CryptoKeyOKP::importJwk):
(WebCore::CryptoKeyOKP::supportsNamedCurve):
(WebCore::CryptoKeyOKP::isPlatformSupportedCurve): Deleted.
* Source/WebCore/crypto/keys/CryptoKeyOKP.h:

Canonical link: <a href="https://commits.webkit.org/286864@main">https://commits.webkit.org/286864@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/665bd6783b5cdd7a25a48d2fb21dc05dd2d1daa6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77246 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56281 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81819 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28528 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79363 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4577 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60518 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18557 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80313 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50484 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66299 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40818 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47887 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23797 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26851 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68998 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24131 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83234 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4626 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3121 "Found 1 new test failure: webrtc/vp8-then-h264.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68786 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4782 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66271 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68042 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17008 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12033 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10133 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4573 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7388 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4592 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8027 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6351 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nmahendru commented on Source/WebCore/crypto/keys/CryptoKeyOKP.h:
```diff
@@ -37,7 +37,7 @@ class CryptoKeyOKP final : public CryptoKey {
     using KeyMaterial = Vector<uint8_t>;
 
-    enum class NamedCurve {
+    enum class NamedCurve : bool {
```
This might add more types like `x448` and `ed448`. `uint8_t` should work.

szewai replied:
For IPC serialization, we prefer using `bool` for enum class that only contains two values as we could validate the values can only be 0 or 1 on receiver side. If we adds the values later, we could update it to uint8_t.

nmahendru commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -691,7 +691,7 @@ static unsigned countUsages(CryptoKeyUsageBitmap usages)
 }
 
-enum class CryptoKeyOKPOpNameTag {
+enum class CryptoKeyOKPOpNameTag : bool {
```
uint8_t as this will most likely need more than 2 values in future. `x448` and `ed448`

szewai replied:
Ditto; we could extend it as needed later.

pvollan reviewed (approved):
R=me.

szewai committed:
Specify underlying type for enum classes in CryptoKey
https://bugs.webkit.org/show_bug.cgi?id=283381
rdar://140231507

Reviewed by Per Arne Vollan.

Also, remove paramter in isPlatformSupportedCurve() and rename it since the return value does not rely on the NamedCurve
value.

* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
* Source/WebCore/crypto/CryptoKey.h:
* Source/WebCore/crypto/CryptoKeyUsage.h:
* Source/WebCore/crypto/SubtleCrypto.h:
* Source/WebCore/crypto/cocoa/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::supportsNamedCurve):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::isPlatformSupportedCurve): Deleted.
* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::CryptoKeyOKP::supportsNamedCurve):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::isPlatformSupportedCurve): Deleted.
* Source/WebCore/crypto/keys/CryptoKeyEC.h:
* Source/WebCore/crypto/keys/CryptoKeyOKP.cpp:
(WebCore::CryptoKeyOKP::generatePair):
(WebCore::CryptoKeyOKP::importRaw):
(WebCore::CryptoKeyOKP::importJwk):
(WebCore::CryptoKeyOKP::supportsNamedCurve):
(WebCore::CryptoKeyOKP::isPlatformSupportedCurve): Deleted.
* Source/WebCore/crypto/keys/CryptoKeyOKP.h:

Canonical link: https://commits.webkit.org/286864@main
